youtube 获取不到1080p的视频

用you-get可以-i url 发现1080p的视频
同一个地址 annie -i url就发现不了

url: "https://www.youtube.com/watch?v=ifn5noo2cYQ"
命令: 
annie -i "https://www.youtube.com/watch?v=ifn5noo2cYQ"
you-get -i "https://www.youtube.com/watch?v=ifn5noo2cYQ"

添加点信息，win10上无法使用annie下载u2b：
这是用 proxifier代理，得到的信息，后面大段重复就没copy了
```
[time-stamp:25] annie.exe *64 - www.youtube.com:443 open through proxy 127.0.0.1:1080 SOCKS5
[time-stamp:26] annie.exe *64 - r6---sn-n4v7sn7y.googlevideo.com:443 open through proxy 127.0.0.1:1080 SOCKS5
[time-stamp:27] annie.exe *64 - r6---sn-n4v7sn7y.googlevideo.com:443 close, 1362 bytes (1.33 KB) sent, 7297 bytes (7.12 KB) received, lifetime 00:01
[time-stamp:27] annie.exe *64 - r6---sn-n4v7sn7y.googlevideo.com:443 open through proxy 127.0.0.1:1080 SOCKS5
[time-stamp:36] chrome.exe *64 - webim.tim.qq.com:443 close, 2435 bytes (2.37 KB) sent, 4999 bytes (4.88 KB) received, lifetime 00:28
[time-stamp:37] annie.exe *64 - r6---sn-n4v7sn7y.googlevideo.com:443 close, 205 bytes sent, 0 bytes received, lifetime 00:10
[time-stamp:37] chrome.exe *64 - webim.tim.qq.com:443 open through proxy 127.0.0.1:1080 SOCKS5
[time-stamp:37] annie.exe *64 - www.youtube.com:443 close, 1644 bytes (1.60 KB) sent, 77316 bytes (75.5 KB) received, lifetime 00:12
[time-stamp:38] wv2ray.exe *64 - 127.8.0.2:5228 error: Cannot connect to placeholder (fake) IP address. It's recomended to restart the application.
[time-stamp:39] wv2ray.exe *64 - 127.8.0.2:5228 error: Cannot connect to placeholder (fake) IP address. It's recomended to restart the application.
[time-stamp:40] wv2ray.exe *64 - 127.8.0.2:5228 error: Cannot connect to placeholder (fake) IP address. It's recomended to restart the application
```

这是控制台输出：
```
# annie.exe https://www.youtube.com/watch?v=w7CisyOOB1o
2018/05/12 00:11:37 https://r6---sn-n4v7sn7y.googlevideo.com/videoplayback?dur=0.000&mv=m&mt=1523518417&ms=au%2Crdu&mn=sn-n4v7sn7y%2Csn-n4v7knls&signature=765476795155C9DC2ECE5EF97C76E0D6A6F6A4AE.3815CE22F767F1C0848EDD4BD1E94CFFEF3AD35F&mm=31%2C29&itag=43&gir=yes&clen=27098181&mime=video%2Fwebm&fvip=1&ipbits=0&requiressl=yes&ratebypass=yes&pl=19&initcwndbps=766250&source=youtube&sparams=clen%2Cdur%2Cei%2Cgir%2Cid%2Cinitcwndbps%2Cip%2Cipbits%2Citag%2Clmt%2Cmime%2Cmm%2Cmn%2Cms%2Cmv%2Cpl%2Cratebypass%2Crequiressl%2Csource%2Cexpire&c=WEB&expire=1523540066&id=o-AKVXItX5-Pp40HFYDup3jGqOHJjX_v_cIMEepeY2zr5f&lmt=1514276676144139&ip=137.175.66.78&key=yt6&ei=AgzPWu2YC8W--AP56IKoAQ
panic: Get https://r6---sn-n4v7sn7y.googlevideo.com/videoplayback?dur=0.000&mv=m&mt=1523518417&ms=au%2Crdu&mn=sn-n4v7sn7y%2Csn-n4v7knls&signature=765476795155C9DC2ECE5EF97C76E0D6A6F6A4AE.3815CE22F767F1C0848EDD4BD1E94CFFEF3AD35F&mm=31%2C29&itag=43&gir=yes&clen=27098181&mime=video%2Fwebm&fvip=1&ipbits=0&requiressl=yes&ratebypass=yes&pl=19&initcwndbps=766250&source=youtube&sparams=clen%2Cdur%2Cei%2Cgir%2Cid%2Cinitcwndbps%2Cip%2Cipbits%2Citag%2Clmt%2Cmime%2Cmm%2Cmn%2Cms%2Cmv%2Cpl%2Cratebypass%2Crequiressl%2Csource%2Cexpire&c=WEB&expire=1523540066&id=o-AKVXItX5-Pp40HFYDup3jGqOHJjX_v_cIMEepeY2zr5f&lmt=1514276676144139&ip=137.175.66.78&key=yt6&ei=AgzPWu2YC8W--AP56IKoAQ: net/http: TLS handshake timeout

goroutine 1 [running]:
github.com/iawia002/annie/request.Request(0x70a88c, 0x3, 0xc0424042c0, 0x293, 0x0, 0x0, 0xc04217f970, 0x20)
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/request/request.go:87 +0xb7f
github.com/iawia002/annie/request.Headers(0xc0424042c0, 0x293, 0xc04206e030, 0x2b, 0x0)
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/request/request.go:131 +0x127
github.com/iawia002/annie/request.Size(0xc0424042c0, 0x293, 0xc04206e030, 0x2b, 0x29)
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/request/request.go:138 +0x54
github.com/iawia002/annie/extractors/youtube.youtubeDownload(0xc04206e030, 0x2b, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/youtube.go:113 +0x49d
github.com/iawia002/annie/extractors/youtube.Download(0xc04206e030, 0x2b)
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/youtube.go:65 +0x30b
main.main()
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/main.go:67 +0x40e
```

--- a/downloader/downloader.go
+++ b/downloader/downloader.go
@@ -2,6 +2,7 @@ package downloader
 
 import (
 	"bufio"
+	"bytes"
 	"encoding/json"
 	"fmt"
 	"io"
@@ -223,22 +224,40 @@ func (v VideoData) Download(refer string) {
 		return
 	}
 	// merge
-	// write ffmpeg input file list
 	mergeFile := title + ".txt" // merge list file should be in the current directory
-	file, _ := os.Create(mergeFile)
-	for _, part := range parts {
-		file.Write([]byte(fmt.Sprintf("file '%s'\n", part)))
-	}
-
 	filePath := utils.FilePath(title, "mp4", false)
 	fmt.Printf("Merging video parts into %s\n", filePath)
-	cmd := exec.Command(
-		"ffmpeg", "-y", "-f", "concat", "-safe", "-1",
-		"-i", mergeFile, "-c", "copy", "-bsf:a", "aac_adtstoasc", filePath,
-	)
+	var cmd *exec.Cmd
+	if strings.Contains(v.Site, "youtube") {
+		// merge audio and video
+		cmds := []string{
+			"-y",
+		}
+		for _, part := range parts {
+			cmds = append(cmds, "-i", part)
+		}
+		cmds = append(
+			cmds, "-c:v", "copy", "-c:a", "aac", "-strict", "experimental",
+			filePath,
+		)
+		cmd = exec.Command("ffmpeg", cmds...)
+	} else {
+		// write ffmpeg input file list
+		file, _ := os.Create(mergeFile)
+		for _, part := range parts {
+			file.Write([]byte(fmt.Sprintf("file '%s'\n", part)))
+		}
+
+		cmd = exec.Command(
+			"ffmpeg", "-y", "-f", "concat", "-safe", "-1",
+			"-i", mergeFile, "-c", "copy", "-bsf:a", "aac_adtstoasc", filePath,
+		)
+	}
+	var stderr bytes.Buffer
+	cmd.Stderr = &stderr
 	err := cmd.Run()
 	if err != nil {
-		log.Fatal(err)
+		log.Fatal(fmt.Sprint(err) + "\n" + stderr.String())
 	}
 	// remove parts
 	os.Remove(mergeFile)

--- a/extractors/douyin.go
+++ b/extractors/douyin.go
@@ -13,8 +13,7 @@ type douyinVideoURLData struct {
 }
 
 type douyinVideoData struct {
-	PlayAddr     douyinVideoURLData `json:"play_addr"`
-	RealPlayAddr string             `json:"real_play_addr"`
+	PlayAddr douyinVideoURLData `json:"play_addr"`
 }
 
 type douyinData struct {
@@ -29,9 +28,10 @@ func Douyin(url string) downloader.VideoData {
 	var dataDict douyinData
 	json.Unmarshal([]byte(vData), &dataDict)
 
-	size := request.Size(dataDict.Video.RealPlayAddr, url)
+	realURL := dataDict.Video.PlayAddr.URLList[0]
+	size := request.Size(realURL, url)
 	urlData := downloader.URLData{
-		URL:  dataDict.Video.RealPlayAddr,
+		URL:  realURL,
 		Size: size,
 		Ext:  "mp4",
 	}

--- a/extractors/weibo.go
+++ b/extractors/weibo.go
@@ -16,7 +16,9 @@ func Weibo(url string) downloader.VideoData {
 	}
 	html := request.Get(url, url)
 	title := utils.MatchOneOf(html, `"content2": "(.+?)"`)[1]
-	realURL := utils.MatchOneOf(html, `"stream_url_hd": "(.+?)"`)[1]
+	realURL := utils.MatchOneOf(
+		html, `"stream_url_hd": "(.+?)"`, `"stream_url": "(.+?)"`,
+	)[1]
 	size := request.Size(realURL, url)
 	urlData := downloader.URLData{
 		URL:  realURL,

--- a/extractors/weibo_test.go
+++ b/extractors/weibo_test.go
@@ -29,6 +29,14 @@ func TestWeibo(t *testing.T) {
 				Size:  438757,
 			},
 		},
+		{
+			name: "normal test",
+			args: test.Args{
+				URL:   "https://m.weibo.cn/status/4226449584043915",
+				Title: "荷兰弟根本不知道跟谁演的对手戏 via@谷大白话",
+				Size:  30429685,
+			},
+		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {

--- a/extractors/youku_test.go
+++ b/extractors/youku_test.go
@@ -16,10 +16,10 @@ func TestYouku(t *testing.T) {
 		{
 			name: "normal test",
 			args: test.Args{
-				URL:     "http://v.youku.com/v_show/id_XMzMzMDk5MzcyNA==.html",
-				Title:   "鲜榨时尚：开口跪Jessie J歌声和造型双在线",
-				Size:    35174132,
-				Quality: "mp4hd3v2 1080x1920",
+				URL:     "http://v.youku.com/v_show/id_XMzUzMjE3NDczNg==.html",
+				Title:   "车事儿：智能汽车已经不在遥远 东风风光iX5发布",
+				Size:    45185427,
+				Quality: "mp4hd3v2 1920x1080",
 			},
 		},
 		{

--- a/extractors/youtube/youtube.go
+++ b/extractors/youtube/youtube.go
@@ -165,6 +165,10 @@ func extractVideoURLS(streams []string, referer, assest string) map[string]downl
 	format["default"] = format[bestQualityItag]
 	delete(format, bestQualityItag)
 
+	// Unlike `url_encoded_fmt_stream_map`, all videos in `adaptive_fmts` have no sound,
+	// we need download video and audio both and then merge them.
+	// Another problem is that even if we add `ratebypass=yes`, the download speed still slow sometimes.
+
 	// All videos here have no sound and need to be added separately
 	for itag, f := range format {
 		if strings.Contains(f.Quality, "video/") {

--- a/extractors/youtube/youtube.go
+++ b/extractors/youtube/youtube.go
@@ -14,9 +14,9 @@ import (
 )
 
 type args struct {
-	Title  string `json:"title"`
-	Stream string `json:"url_encoded_fmt_stream_map"`
-	Audio  string `json:"adaptive_fmts"`
+	Title string `json:"title"`
+	// Stream string `json:"url_encoded_fmt_stream_map"`
+	Stream string `json:"adaptive_fmts"`
 }
 
 type assets struct {
@@ -104,26 +104,6 @@ func youtubeDownload(uri string) downloader.VideoData {
 
 	format := extractVideoURLS(streams, uri, youtube.Assets.JS)
 
-	// Audio only file
-	for _, s := range strings.Split(youtube.Args.Audio, ",") {
-		stream, _ := url.ParseQuery(s)
-		if strings.HasPrefix(stream.Get("type"), "audio/mp4") {
-			audioURL := genSignedURL(stream.Get("url"), stream, youtube.Assets.JS) + "&ratebypass=yes"
-			size := request.Size(audioURL, uri)
-			urlData := downloader.URLData{
-				URL:  audioURL,
-				Size: size,
-				Ext:  "m4a",
-			}
-			format["audio"] = downloader.FormatData{
-				URLs:    []downloader.URLData{urlData},
-				Size:    size,
-				Quality: stream.Get("type"),
-			}
-			break
-		}
-	}
-
 	extractedData := downloader.VideoData{
 		Site:    "YouTube youtube.com",
 		Title:   utils.FileName(title),
@@ -135,6 +115,8 @@ func youtubeDownload(uri string) downloader.VideoData {
 }
 
 func extractVideoURLS(streams []string, referer, assest string) map[string]downloader.FormatData {
+	var ext string
+	var audio downloader.URLData
 	format := map[string]downloader.FormatData{}
 
 	bestQualityURL, _ := url.ParseQuery(streams[0])
@@ -143,20 +125,35 @@ func extractVideoURLS(streams []string, referer, assest string) map[string]downl
 	for _, s := range streams {
 		stream, _ := url.ParseQuery(s)
 		itag := stream.Get("itag")
+		streamType := stream.Get("type")
+		isAudio := strings.HasPrefix(streamType, "audio/mp4")
 
-		if !utils.ShouldExtract(itag, bestQualityItag) {
+		if !isAudio && !utils.ShouldExtract(itag, bestQualityItag) {
 			continue
 		}
 
-		quality := stream.Get("quality")
-		ext := utils.MatchOneOf(stream.Get("type"), `video/(\w+);`)[1]
-		streamURL := stream.Get("url")
-		realURL := genSignedURL(streamURL, stream, assest)
+		quality := stream.Get("quality_label")
+		if quality != "" {
+			quality = fmt.Sprintf("%s %s", quality, streamType)
+		} else {
+			quality = streamType
+		}
+		if isAudio {
+			// audio file use m4a extension
+			ext = "m4a"
+		} else {
+			ext = utils.MatchOneOf(streamType, `(\w+)/(\w+);`)[2]
+		}
+		realURL := genSignedURL(stream.Get("url"), stream, assest) + "&ratebypass=yes"
 		size := request.Size(realURL, referer)
 		urlData := downloader.URLData{
 			URL:  realURL,
 			Size: size,
 			Ext:  ext,
+		}
+		if ext == "m4a" {
+			// Audio data for merging with video
+			audio = urlData
 		}
 		format[itag] = downloader.FormatData{
 			URLs:    []downloader.URLData{urlData},
@@ -167,5 +164,14 @@ func extractVideoURLS(streams []string, referer, assest string) map[string]downl
 
 	format["default"] = format[bestQualityItag]
 	delete(format, bestQualityItag)
+
+	// All videos here have no sound and need to be added separately
+	for itag, f := range format {
+		if strings.Contains(f.Quality, "video/") {
+			f.Size += audio.Size
+			f.URLs = append(format[itag].URLs, audio)
+			format[itag] = f
+		}
+	}
 	return format
 }

--- a/extractors/youtube/youtube_test.go
+++ b/extractors/youtube/youtube_test.go
@@ -19,8 +19,8 @@ func TestYoutube(t *testing.T) {
 			args: test.Args{
 				URL:     "https://www.youtube.com/watch?v=Gnbch2osEeo",
 				Title:   "Multifandom Mashup 2017",
-				Size:    60785404,
-				Quality: "hd720",
+				Size:    60808458,
+				Quality: `720p video/mp4; codecs="avc1.4d401f"`,
 			},
 		},
 		{
@@ -28,8 +28,8 @@ func TestYoutube(t *testing.T) {
 			args: test.Args{
 				URL:     "https://youtu.be/z8eFzkfto2w",
 				Title:   "Circle Of Love - Rudy Mancuso",
-				Size:    27183162,
-				Quality: "hd720",
+				Size:    37244990,
+				Quality: `1080p video/mp4; codecs="avc1.640028"`,
 			},
 		},
 		{
@@ -37,8 +37,8 @@ func TestYoutube(t *testing.T) {
 			args: test.Args{
 				URL:     "https://www.youtube.com/watch?v=ASPku-eAZYs",
 				Title:   "怪獸與葛林戴華德的罪行 - HD首版電影預告大首播 (Fantastic Beasts：The Crimes of Grindelwald)",
-				Size:    18330678,
-				Quality: "hd720",
+				Size:    31655118,
+				Quality: `1080p video/mp4; codecs="avc1.640028"`,
 			},
 		},
 		{
@@ -46,8 +46,8 @@ func TestYoutube(t *testing.T) {
 			args: test.Args{
 				URL:     "https://www.youtube.com/watch?v=x2nKigmfzLQ&list=PLfyyPldyYFapPfxZOay2AoCCaNE1Ezd4_",
 				Title:   "【仔細看看】《羅根》黑白版-漫畫電影新高峰?! - 超粒方",
-				Size:    108941976,
-				Quality: "hd720",
+				Size:    180227347,
+				Quality: `1080p video/mp4; codecs="avc1.640028"`,
 			},
 			playlist: true,
 		},

